lib/galaxy/datatypes/molecules.py

```diff
@@ -92,26 +92,53 @@ def sniff(self, filename):
         """
         Try to guess if the file is a SDF2 file.
 
+        An SDfile (structure-data file) can contain multiple compounds.
+
+        Each compound starts with a block in V2000 or V3000 molfile format,
+        which ends with a line equal to 'M  END'.
+        This is followed by a non-structural data block, which ends with a line
+        equal to '$$$$'.
+
         >>> from galaxy.datatypes.sniff import get_test_fname
         >>> fname = get_test_fname('drugbank_drugs.sdf')
         >>> SDF().sniff(fname)
         True
 
-        >>> fname = get_test_fname('drugbank_drugs.cml')
+        >>> fname = get_test_fname('github88.v3k.sdf')
+        >>> SDF().sniff(fname)
+        True
+
+        >>> fname = get_test_fname('chebi_57262.v3k.mol')
         >>> SDF().sniff(fname)
         False
         """
-        counter = count_special_lines("^M\s*END", filename) + count_special_lines("^\$\$\$\$", filename)
-        if counter > 0 and counter % 2 == 0:
-            return True
-        else:
-            return False
+        m_end_found = False
+        limit = 10000
+        idx = 0
+        with open(filename) as in_file:
+            for line in in_file:
```

Comment:
We really need to stop doing this - this line can consume unbounded memory on relatively sane user supplied inputs. I guess there are a lot of instances of it though and I'm not finding my previous advice on fixing it. (I believe the right thing to do is pick a maximum line size and use readlines directly instead of the iterator).

Reply:
xref. https://github.com/galaxyproject/galaxy/issues/4319